src/main/java/org/broadinstitute/hellbender/tools/copynumber/utils/segmentation/KernelSegmenter.java

```diff
@@ -0,0 +1,526 @@
+     * Various recurrence relations are used to calculate costs iteratively.
+     * @param reducedObservationMatrix      N x p matrix of projected observations, where N is the number of data points
+     *                                      and p is the dimension of the low-rank approximation of the kernel matrix;
+     *                                      this is the Z matrix described in the text preceding Eq. 14, ibid
+     * @param kernelApproximationDiagonal   N diagonal terms of the low-rank approximation to the kernel matrix
+     * @param windowSize                    number of points to include in either flanking segment when calculating cost
+     */
+    private static double[] calculateWindowCosts(final RealMatrix reducedObservationMatrix,
+                                                 final double[] kernelApproximationDiagonal,
+                                                 final int windowSize) {
+        final int N = reducedObservationMatrix.getRowDimension();
+        final int p = reducedObservationMatrix.getColumnDimension();
+
+        //initialize indices of the boundaries of the two flanking segments, wrapping around to beginning of data if necessary
+        int center = 0;
+        int start = (center - windowSize + 1 + N) % N;
+        int end = (center + windowSize) % N;
+
+        //initialize costs of flanking segments and total segment
+        final Cost leftCost = calculateSegmentCost(start, center, reducedObservationMatrix, kernelApproximationDiagonal);
+        final Cost rightCost = calculateSegmentCost(center + 1, end, reducedObservationMatrix, kernelApproximationDiagonal);
+        final Cost totalCost = calculateSegmentCost(start, end, reducedObservationMatrix, kernelApproximationDiagonal);
+
+        //initialize quantities for recurrence
+        double leftD = leftCost.D;
+        final double[] leftW = Arrays.copyOf(leftCost.W, p);
+        double leftV = leftCost.V;
+        double leftC = leftCost.C;
+
+        double rightD = rightCost.D;
+        final double[] rightW = Arrays.copyOf(rightCost.W, p);
+        double rightV = rightCost.V;
+        double rightC = rightCost.C;
+
+        double totalD = totalCost.D;
+        final double[] totalW = Arrays.copyOf(totalCost.W, p);
+        double totalV = totalCost.V;
+        double totalC = totalCost.C;
+
+        final double[] windowCosts = new double[N];
+        windowCosts[center] = leftC + rightC - totalC;
+
+        double ZdotW;
+        final double windowSizeReciprocal = 1. / windowSize;
+
+        //slide segments along data and use recurrence relations to iteratively update costs
+        for (center = 0; center < N; center++) {
+            final int centerNext = (center + 1) % N;
+            final int endNext = (end + 1) % N;
+
+            //update quantities in left segment
+            leftD -= kernelApproximationDiagonal[start];
+            ZdotW = 0.;
+            for (int j = 0; j < p; j++) {
```

Comment:
Can you extract a method for this four-line motif?

Reply:
I thought about it, but decided against it because such a method needs to either return `ZdotW` and modify the `W` costs as a side effect or return a pair of quantities.  There are also enough variations on the motif to require parameterization of the method.  But it does shorten up the code a bit, so done!